File: src/main/java/net/igneo/icv/mixin/EnchantmentScreenMixin.java
package net.igneo.icv.mixin;

import com.google.common.collect.Lists;
import com.mojang.blaze3d.platform.Lighting;
import com.mojang.blaze3d.vertex.VertexConsumer;
import com.mojang.math.Axis;
import net.igneo.icv.ICV;
import net.igneo.icv.event.ModEvents;
import net.minecraft.ChatFormatting;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;
import net.minecraft.client.gui.screens.inventory.EnchantmentNames;
import net.minecraft.client.gui.screens.inventory.EnchantmentScreen;
import net.minecraft.client.model.BookModel;
import net.minecraft.client.renderer.texture.OverlayTexture;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.network.chat.CommonComponents;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.FormattedText;
import net.minecraft.network.chat.MutableComponent;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.Mth;
import net.minecraft.world.SimpleContainer;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.inventory.EnchantmentMenu;
import net.minecraft.world.item.enchantment.Enchantment;
import org.joml.Matrix4f;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Overwrite;
import org.spongepowered.asm.mixin.Shadow;

import java.util.List;

@Mixin(value = EnchantmentScreen.class,priority = 999999999)
public class EnchantmentScreenMixin extends AbstractContainerScreen<EnchantmentMenu> {
    @Shadow
    private static final ResourceLocation ENCHANTING_TABLE_LOCATION = new ResourceLocation(ICV.MOD_ID,"textures/gui/container/enchanting_table.png");
    @Shadow
    private static final ResourceLocation ENCHANTING_BOOK_LOCATION = new ResourceLocation("textures/entity/enchanting_table_book.png");
    @Shadow
    private BookModel bookModel;
    @Shadow
    public float flip;
    @Shadow
    public float oFlip;
    @Shadow
    public float flipT;
    @Shadow
    public float flipA;
    @Shadow
    public float open;
    @Shadow
    public float oOpen;
    public EnchantmentScreenMixin(EnchantmentMenu pMenu, Inventory pPlayerInventory, Component pTitle) {
        super(pMenu, pPlayerInventory, pTitle);
    }

    @Override
    public boolean mouseScrolled(double pMouseX, double pMouseY, double pDelta) {
        if (-pDelta > 0) {
            this.minecraft.gameMode.handleInventoryButtonClick((this.menu).containerId, -1);
        } else {
            this.minecraft.gameMode.handleInventoryButtonClick((this.menu).containerId, -2);
        }
        return super.mouseScrolled(pMouseX, pMouseY, pDelta);
    }

    /**
     * @author Igneo220
     * @reason enchantment revamp
     */
    @Overwrite
    public boolean mouseClicked(double pMouseX, double pMouseY, int pButton) {
        int i = (this.width - this.imageWidth) / 2;
        int j = (this.height - this.imageHeight) / 2;

        for(int k = 0; k < 3; ++k) {
            double d0 = pMouseX - (double)(i + 60);
            double d1 = pMouseY - (double)(j + 14 + 19 * k);
            int l = k + ModEvents.enchShift;
            if (d0 >= 0.0D && d1 >= 0.0D && d0 < 108.0D && d1 < 19.0D && this.menu.clickMenuButton(this.minecraft.player, l)) {
                this.minecraft.gameMode.handleInventoryButtonClick((this.menu).containerId, l);
                return true;
            }
        }

        return super.mouseClicked(pMouseX, pMouseY, pButton);
    }

    /**
     * @author Igneo220
     * @reason changing enchantment table
     */
    @Overwrite
    protected void renderBg(GuiGraphics pGuiGraphics, float pPartialTick, int pMouseX, int pMouseY) {
        int i = (this.width - this.imageWidth) / 2;
        int j = (this.height - this.imageHeight) / 2;
        //BIG GRAY PART
        pGuiGraphics.blit(ENCHANTING_TABLE_LOCATION, i, j, 0, 0, this.imageWidth, this.imageHeight);
        //da book
        this.renderBook(pGuiGraphics, i, j, pPartialTick);
        //erm, not sure
        //EnchantmentNames.getInstance().initSeed((long)this.menu.getEnchantmentSeed());
        int k = this.menu.getGoldCount();

        for(int l = 0; l < 3; ++l) {
            int i1 = i + 60;
            int j1 = i1 + 20;
            int k1 = 1;
            if ((this.menu).enchantClue[l] == -1 || k1 == 0) {
                pGuiGraphics.blit(ENCHANTING_TABLE_LOCATION, i1, j + 14 + 19 * l, 0, 185, 108, 19);
            } else {
                String s = "" + k1;
                int l1 = 86 - this.font.width(s);
                Enchantment enchantment = Enchantment.byId((this.menu).enchantClue[l]);
                FormattedText formattedtext = (Component.translatable("%s", enchantment == null ? "" : enchantment.getFullname((this.menu).levelClue[l])).withStyle(ChatFormatting.BLACK));
                int i2 = 6839882;
                if (((k < l + 1 || this.minecraft.player.experienceLevel < k1) && !this.minecraft.player.getAbilities().instabuild) || this.menu.enchantClue[l] == -1) { // Forge: render buttons as disabled when enchantable but enchantability not met on lower levels
                    pGuiGraphics.blit(ENCHANTING_TABLE_LOCATION, i1, j + 14 + 19 * l, 0, 185, 108, 19);
                    //pGuiGraphics.blit(ENCHANTING_TABLE_LOCATION, i1 + 1, j + 15 + 19 * l, 16 * l, 239, 16, 16);
                    //pGuiGraphics.drawWordWrap(this.font, formattedtext, j1, j + 16 + 19 * l, l1, (i2 & 16711422) >> 1);
                    i2 = 4226832;
                } else {
                    int j2 = pMouseX - (i + 60);
                    int k2 = pMouseY - (j + 14 + 19 * l);
                    if (j2 >= 0 && k2 >= 0 && j2 < 108 && k2 < 19) {
                        pGuiGraphics.blit(ENCHANTING_TABLE_LOCATION, i1, j + 14 + 19 * l, 0, 204, 108, 19);
                        i2 = 16777088;
                    } else {
                        pGuiGraphics.blit(ENCHANTING_TABLE_LOCATION, i1, j + 14 + 19 * l, 0, 166, 108, 19);
                    }
                    //this.font.drawInBatch(pose, label, labelX, labelY, 0x404040);
                    //pGuiGraphics.drawString()
                    //this.font.drawInBatch(formattedtext.getString(),j1,(j + 16 + 19 * l) + 3,99999999,true, new Matrix4f(1),1);
                    //ORB
                    pGuiGraphics.blit(ENCHANTING_TABLE_LOCATION, i1 + 1, j + 15 + 19 * l, 16 * l, 223, 16, 16);
                    //formattedtext = FormattedText.EMPTY;
                    //pGuiGraphics.drawWordWrap(this.font, formattedtext, j1, (j + 16 + 19 * l) + 3, l1, 999999999);
                    pGuiGraphics.drawString(this.font,formattedtext.getString(),j1,(j + 16 + 19 * l) + 3,-587202561,true);
                    i2 = 8453920;
                }

                //pGuiGraphics.drawString(this.font, s, j1 + 86 - this.font.width(s), j + 16 + 19 * l + 7, i2);
            }
        }

    }

    @Shadow
    private void renderBook(GuiGraphics pGuiGraphics, int pX, int pY, float pPartialTick) {
        float f = Mth.lerp(pPartialTick, this.oOpen, this.open);
        float f1 = Mth.lerp(pPartialTick, this.oFlip, this.flip);
        Lighting.setupForEntityInInventory();
        pGuiGraphics.pose().pushPose();
        pGuiGraphics.pose().translate((float)pX + 33.0F, (float)pY + 31.0F, 100.0F);
        float f2 = 40.0F;
        pGuiGraphics.pose().scale(-40.0F, 40.0F, 40.0F);
        pGuiGraphics.pose().mulPose(Axis.XP.rotationDegrees(25.0F));
        pGuiGraphics.pose().translate((1.0F - f) * 0.2F, (1.0F - f) * 0.1F, (1.0F - f) * 0.25F);
        float f3 = -(1.0F - f) * 90.0F - 90.0F;
        pGuiGraphics.pose().mulPose(Axis.YP.rotationDegrees(f3));
        pGuiGraphics.pose().mulPose(Axis.XP.rotationDegrees(180.0F));
        float f4 = Mth.clamp(Mth.frac(f1 + 0.25F) * 1.6F - 0.3F, 0.0F, 1.0F);
        float f5 = Mth.clamp(Mth.frac(f1 + 0.75F) * 1.6F - 0.3F, 0.0F, 1.0F);
        this.bookModel.setupAnim(0.0F, f4, f5, f);
        VertexConsumer vertexconsumer = pGuiGraphics.bufferSource().getBuffer(this.bookModel.renderType(ENCHANTING_BOOK_LOCATION));
        this.bookModel.renderToBuffer(pGuiGraphics.pose(), vertexconsumer, 15728880, OverlayTexture.NO_OVERLAY, 1.0F, 1.0F, 1.0F, 1.0F);
        pGuiGraphics.flush();
        pGuiGraphics.pose().popPose();
        Lighting.setupFor3DItems();
    }


    /**
     * @author Igneo220
     * @reason changing enchantment table
     */
    @Overwrite
    public void render(GuiGraphics pGuiGraphics, int pMouseX, int pMouseY, float pPartialTick) {
        pPartialTick = this.minecraft.getFrameTime();
        this.renderBackground(pGuiGraphics);
        super.render(pGuiGraphics, pMouseX, pMouseY, pPartialTick);
        this.renderTooltip(pGuiGraphics, pMouseX, pMouseY);
        boolean flag = this.minecraft.player.getAbilities().instabuild;
        int i = this.menu.getGoldCount();

        for(int j = 0; j < 3; ++j) {
            int k = (this.menu).costs[j];
            Enchantment enchantment = Enchantment.byId((this.menu).enchantClue[j]);
            int l = (this.menu).levelClue[j];
            int i1 = j + 1;
            if (this.isHovering(60, 14 + 19 * j, 108, 17, (double)pMouseX, (double)pMouseY) && k > 0) {
                List<Component> list = Lists.newArrayList();
                //NECESSARY FOR HOVER TEXT
                list.add((Component.translatable("enchantment." + BuiltInRegistries.ENCHANTMENT.getKey(enchantment).toString().replace(":",".") + ".desc", enchantment == null ? "" : enchantment.getFullname(l))).withStyle(ChatFormatting.WHITE));
                if (enchantment == null) {
                    //ANNOYING WEIRD STUFF IT SHOWS WHEN NOTHINGS THERE
                    //list.add(Component.literal(""));
                    //list.add(Component.translatable("forge.container.enchant.limitedEnchantability").withStyle(ChatFormatting.RED));
                } else if (!flag) {
                    //no idea what this does, try uncommenting it if things break
                    //list.add(CommonComponents.EMPTY);
                    if (this.minecraft.player.experienceLevel < k) {
                        //no idea what this does either
                        //list.add(Component.translatable("container.enchant.level.requirement", (this.menu).costs[j]).withStyle(ChatFormatting.RED));
                    } else {
                        MutableComponent mutablecomponent;
                        if (i1 == 1) {
                            mutablecomponent = Component.translatable("container.enchant.lapis.one");
                        } else {
                            mutablecomponent = Component.translatable("container.enchant.lapis.many", i1);
                        }

                        //list.add(mutablecomponent.withStyle(i >= i1 ? ChatFormatting.GRAY : ChatFormatting.RED));
                        MutableComponent mutablecomponent1;
                        if (i1 == 1) {
                            mutablecomponent1 = Component.translatable("container.enchant.level.one");
                        } else {
                            mutablecomponent1 = Component.translatable("container.enchant.level.many", i1);
                        }

                        //list.add(mutablecomponent1.withStyle(ChatFormatting.GRAY));
                    }
                }

                //hello, I am the floater thingie
                //i say the stuff
                FormattedText formattedtext = (Component.translatable("%s", enchantment == null ? "" : enchantment.getFullname((this.menu).levelClue[l])).withStyle(ChatFormatting.BLACK));
                pGuiGraphics.renderComponentTooltip(this.font, list, pMouseX, pMouseY);
                break;
            }
        }

    }
}

File: src/main/java/net/igneo/icv/mixin/EnchantmentScreenMixin.kt
@file:Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE") // For Mixin + Forge compatibility

package net.igneo.icv.mixin

import com.google.common.collect.Lists
import com.mojang.blaze3d.platform.Lighting
import com.mojang.blaze3d.vertex.VertexConsumer
import com.mojang.math.Axis
import net.igneo.icv.ICV
import net.minecraft.ChatFormatting
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen
import net.minecraft.client.gui.screens.inventory.EnchantmentScreen
import net.minecraft.client.model.BookModel
import net.minecraft.client.renderer.texture.OverlayTexture
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation
import net.minecraft.util.Mth
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.inventory.EnchantmentMenu
import net.minecraft.world.item.enchantment.Enchantment
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Overwrite
import org.spongepowered.asm.mixin.Shadow

@Mixin(EnchantmentScreen::class)
abstract class EnchantmentScreenMixin(
    menu: EnchantmentMenu,
    inventory: Inventory,
    title: Component
) : AbstractContainerScreen<EnchantmentMenu>(menu, inventory, title) {

    @Shadow private lateinit var bookModel: BookModel
    @Shadow private var flip = 0f
    @Shadow private var oFlip = 0f
    @Shadow private var flipT = 0f
    @Shadow private var flipA = 0f
    @Shadow private var open = 0f
    @Shadow private var oOpen = 0f

    companion object {
        private val TABLE_TEX = ResourceLocation(ICV.MOD_ID, "textures/gui/container/enchanting_table.png")
        private val BOOK_TEX = ResourceLocation("textures/entity/enchanting_table_book.png")
    }

    override fun mouseScrolled(mouseX: Double, mouseY: Double, delta: Double): Boolean {
        minecraft?.gameMode?.handleInventoryButtonClick(menu.containerId, if (delta < 0) -1 else -2)
        return super.mouseScrolled(mouseX, mouseY, delta)
    }

    /**
     * @author Igneo220
     * @reason enchantment revamp
     */
    @Overwrite
    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        val left = (width - imageWidth) / 2
        val top = (height - imageHeight) / 2

        for (i in 0..2) {
            val x = mouseX - (left + 60)
            val y = mouseY - (top + 14 + 19 * i)
            if (x in 0.0..108.0 && y in 0.0..19.0 && menu.clickMenuButton(minecraft?.player, i)) {
                minecraft?.gameMode?.handleInventoryButtonClick(menu.containerId, i)
                return true
            }
        }

        return super.mouseClicked(mouseX, mouseY, button)
    }

    /**
     * @author Igneo220
     * @reason changing enchantment table
     */
    @Overwrite
    override fun renderBg(graphics: GuiGraphics, partialTick: Float, mouseX: Int, mouseY: Int) {
        val left = (width - imageWidth) / 2
        val top = (height - imageHeight) / 2

        graphics.blit(TABLE_TEX, left, top, 0, 0, imageWidth, imageHeight)
        renderBook(graphics, left, top, partialTick)

        val gold = menu.goldCount

        for (i in 0..2) {
            val x = left + 60
            val y = top + 14 + 19 * i
            val hasEnchant = menu.enchantClue[i] != -1

            if (!hasEnchant) {
                graphics.blit(TABLE_TEX, x, y, 0, 185, 108, 19)
                continue
            }

            val enchant = Enchantment.byId(menu.enchantClue[i])
            val label = Component.translatable("%s", enchant?.getFullname(menu.levelClue[i]) ?: "").withStyle(ChatFormatting.BLACK)
            val cost = 1
            val lapisX = x + 20
            val lapisY = y + 2
            val highlight = mouseX in x..(x + 107) && mouseY in y..(y + 18)

            when {
                (gold < i + 1 || (minecraft?.player?.experienceLevel ?: 0) < cost) && !minecraft?.player?.abilities?.instabuild!! -> {
                    graphics.blit(TABLE_TEX, x, y, 0, 185, 108, 19)
                }
                highlight -> {
                    graphics.blit(TABLE_TEX, x, y, 0, 204, 108, 19)
                    graphics.blit(TABLE_TEX, x + 1, y + 1, 16 * i, 223, 16, 16)
                    graphics.drawString(font, label.string, lapisX, lapisY + 12, -0x23000001, true)
                }
                else -> {
                    graphics.blit(TABLE_TEX, x, y, 0, 166, 108, 19)
                    graphics.blit(TABLE_TEX, x + 1, y + 1, 16 * i, 223, 16, 16)
                    graphics.drawString(font, label.string, lapisX, lapisY + 12, 8453920, true)
                }
            }
        }
    }

    @Shadow
    private fun renderBook(graphics: GuiGraphics, x: Int, y: Int, partialTick: Float) {
        val openInterp = Mth.lerp(partialTick, oOpen, open)
        val flipInterp = Mth.lerp(partialTick, oFlip, flip)
        Lighting.setupForEntityInInventory()
        graphics.pose().apply {
            pushPose()
            translate(x + 33f, y + 31f, 100f)
            scale(-40f, 40f, 40f)
            mulPose(Axis.XP.rotationDegrees(25f))
            translate((1 - openInterp) * 0.2f, (1 - openInterp) * 0.1f, (1 - openInterp) * 0.25f)
            mulPose(Axis.YP.rotationDegrees(-(1 - openInterp) * 90 - 90))
            mulPose(Axis.XP.rotationDegrees(180f))
        }

        val f4 = Mth.clamp(Mth.frac(flipInterp + 0.25f) * 1.6f - 0.3f, 0f, 1f)
        val f5 = Mth.clamp(Mth.frac(flipInterp + 0.75f) * 1.6f - 0.3f, 0f, 1f)

        bookModel.setupAnim(0f, f4, f5, openInterp)
        val consumer: VertexConsumer = graphics.bufferSource().getBuffer(bookModel.renderType(BOOK_TEX))
        bookModel.renderToBuffer(graphics.pose(), consumer, 15728880, OverlayTexture.NO_OVERLAY, 1f, 1f, 1f, 1f)
        graphics.flush()
        graphics.pose().popPose()
        Lighting.setupFor3DItems()
    }

    /**
     * @author Igneo220
     * @reason render tooltips
     */
    @Overwrite
    override fun render(graphics: GuiGraphics, mouseX: Int, mouseY: Int, partialTick: Float) {
        renderBackground(graphics)
        super.render(graphics, mouseX, mouseY, minecraft!!.frameTime)
        renderTooltip(graphics, mouseX, mouseY)

        val canCheat = minecraft!!.player?.abilities?.instabuild == true

        for (i in 0..2) {
            val x = 60
            val y = 14 + 19 * i
            if (!isHovering(x, y, 108, 17, mouseX.toDouble(), mouseY.toDouble()) || menu.costs[i] <= 0) continue

            val clues = menu.enchantClue
            val levels = menu.levelClue
            val enchant = Enchantment.byId(clues[i])
            val components = Lists.newArrayList<Component>()

            if (enchant != null) {
                val key = BuiltInRegistries.ENCHANTMENT.getKey(enchant).toString().replace(":", ".").replace(" ", "_")
                components.add(Component.translatable("enchantment.$key.desc", enchant.getFullname(levels[i])).withStyle(ChatFormatting.WHITE))
            }

            if (enchant != null && !canCheat) {
                val lapis = Component.translatable("container.enchant.lapis.${if (i + 1 == 1) "one" else "many"}", i + 1)
                val level = Component.translatable("container.enchant.level.${if (i + 1 == 1) "one" else "many"}", i + 1)
                // Add these components to the tooltip if desired
            }

            graphics.renderComponentTooltip(font, components, mouseX, mouseY)
            break
        }
    }
}